bokeh/document.py

```diff
@@ -298,6 +300,12 @@ def __init__(self, **kwargs):
         self._callbacks = {}
         self._session_callbacks = {}
         self._session_context = None
+        self._modules = []
+
+    def __del__(self):
+        for module in self._modules:
+            if module.__name__ in sys.modules:
+                del sys.modules[module.__name__]
```

Comment:
Given this code:
```
 def modify_document(self, doc):
     if self.failed:
         return
     module = self._runner.new_module()
```
 in `bokeh/application/handlers/code.py` I would say it's safe to assume that a module won't be shared across documents. Though, we may have multiple modules referring to one piece of code per document. I'm not sure when or how often `modify_document()` is called, so I'm not sure if it's a problem or not.